projects/noloan/admin/src/main/java/hasadna/noloan/admin/app/SpamRecyclerAdapter.java

```diff
@@ -13,35 +13,66 @@
 import android.widget.TextView;
 import android.widget.Toast;
 
+import java.util.ArrayList;
+
 import hasadna.noloan.common.FirestoreClient;
 import hasadna.noloan.common.SmsMessages;
 import hasadna.noloan.protobuf.SmsProto.SmsMessage;
 
 public class SpamRecyclerAdapter
     extends RecyclerView.Adapter<SpamRecyclerAdapter.RecyclerViewHolder> {
 
-  SmsMessages DbMessages;
+  ArrayList<SmsMessage> messages;
 
   public SpamRecyclerAdapter() {
 
-    DbMessages = SmsMessages.get();
+    messages = new ArrayList<>();
+    SmsMessages dbMessages = SmsMessages.get();
+
+    // maybe add getSuggestions to SmsMessages
+    for (SmsMessage message : dbMessages.getDbMessages()) {
+      if (message.getApproved()) {
+        messages.add(message);
+      }
+    }
+
     Handler handler = new Handler(Looper.getMainLooper());
 
-    DbMessages.setMessagesListener(
+    dbMessages.setMessagesListener(
         new SmsMessages.MessagesListener() {
           @Override
           public void messageAdded(SmsMessage smsMessage) {
-            handler.post(() -> notifyItemInserted(DbMessages.getDbMessages().size()));
+            if (smsMessage.getApproved()) {
+              messages.add(smsMessage);
+            }
+            handler.post(() -> notifyItemInserted(messages.size()));
           }
 
           @Override
           public void messageRemoved(int index, SmsMessage smsMessage) {
-            handler.post(() -> notifyItemRemoved(index));
+            if (smsMessage.getApproved()) {
+              int i = SmsMessages.searchMessage(smsMessage, messages);
+              messages.remove(smsMessage);
+              handler.post(() -> notifyItemRemoved(i));
+            }
           }
 
           @Override
           public void messageModified(int index) {
-            handler.post(() -> notifyItemChanged(index));
+            SmsMessage smsMessage = dbMessages.getDbMessages().get(index);
+            int i = SmsMessages.searchMessage(smsMessage, messages);
+            if (i != -1 && smsMessage.getApproved()) { // message approved and in the list
+              messages.set(i, smsMessage);
+              handler.post(() -> notifyItemChanged(i));
+            } else if (i != -1) // message not approved but in the list
+            {
+              messages.remove(i);
+              handler.post(() -> notifyItemRemoved(i));
+            } else // message approved but not in the list
```

Comment:
Just to understand...

This is a case which an approved message (written in the DB) had been changed to be disapproved - right?

This can happen (at the moment) only if:
1. This change came from some reason - from the DB
2. Coming from the admin app - Do we have an "undo" for `approve` ?

This cannot come from the NoLoan app right?

Reply:
Cool, thanks :+1: